scripts/gen_csr_test.py

```diff
@@ -0,0 +1,220 @@
+"""
+def gen_csr_instr(csr_map, csr_instructions, isa_length, iterations, out):
+  for i in range(iterations):
+    # pick two GPRs at random to act as source and destination registers
+    # for CSR operations
+    source_reg, dest_reg = [f"x{i}" for i in random.sample(range(5, 15), 2)] #support risc-v E extension (16 registers)
+    csr_list = list(csr_map.keys())
+    with open(f"{out}/riscv_csr_test.{i}.S", "w") as csr_test_file:
+      # Write test label
+      csr_test_file.write(f"csr_test:\n")
+      for csr in csr_list:
+        last_csr = csr
+        csr_val, csr_mask = csr_map.get(csr)
+        for op in csr_instructions:
+          for i in range(3):
+            # hex string
+            rand_rs1_val = get_rs1_val(i, isa_length)
+            # I type CSR instruction
+            first_li = ""
+            if op[-1] == "i":
+              imm = rand_rs1_val[-5:]
+              csr_inst = f"\t{op} {dest_reg}, {csr}, 0b{imm.bin}\n"
+            else:
+              first_li = f"\tli {source_reg}, 0x{rand_rs1_val.hex}\n"
+              csr_inst = f"\t{op} {dest_reg}, {csr}, {source_reg}\n"
+              predict_li = f"\tli {source_reg}, {predict_csr_val(op, rand_rs1_val, csr_val, csr_mask)}\n"
+            branch_check = f"\tbne {source_reg}, {dest_reg}, csr_fail\n"
+            # write test sequence to test file
+            csr_test_file.write(first_li)
+            csr_test_file.write(csr_inst)
+            csr_test_file.write(predict_li)
+            csr_test_file.write(branch_check)
+            # Need to hardcode in one last branch check, since the value last written to
+            # CSR has not been tested at this point
+            # Only if we have finished
+            if csr == csr_list[-1] and op == csr_instructions[-1] and i == 2:
+              final_csr_read = f"\tcsrr {dest_reg}, {csr_list[-1]}\n"
+              csrrs_read_mask = bitarray(uint=0, length=isa_length)
+              final_li = f"\tli {source_reg}, {predict_csr_val('csrrs', csrrs_read_mask, csr_val, csr_mask)}\n"
+              final_branch_check = f"\tbne {source_reg}, {dest_reg}, csr_fail\n"
+              # write to test file
+              csr_test_file.write(final_csr_read)
+              csr_test_file.write(final_li)
+              csr_test_file.write(final_branch_check)
+      # Add final signatures to test, indicating to testbench that test has ended
+      gen_csr_test_pass(csr_test_file)
+      gen_csr_test_fail(csr_test_file)
+
+
+# Add argument options
+parser = argparse.ArgumentParser()
+parser.add_argument("--csr_file", type=str, help="The YAML file contating descriptions of all processor supported CSRs")
+parser.add_argument("--rv", type=int, default=32, help="Specify the ISA width, e.g. 32 or 64 or 128")
+parser.add_argument("--num_test", type=int, default=1, help="Specify how many tests to be generated")
```

Comment:
Looks like this is not used

Reply:
num_test is used by gen_csr_inst() to specify how many output files to generate